File: app/src/main/java/ga/forntoh/bableschool/MainActivity.java
package ga.forntoh.bableschool;

import android.annotation.SuppressLint;
import android.os.Bundle;

import com.forntoh.EasyRecyclerView.EasyRecyclerView;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import ga.forntoh.bableschool.adapters.CategoryAdapter;
import ga.forntoh.bableschool.model.Category;
import ga.forntoh.bableschool.network.ApiService;
import ga.forntoh.bableschool.network.RetrofitBuilder;
import ga.forntoh.bableschool.store.MyStores;
import ga.forntoh.bableschool.store.StoreRepository;
import io.reactivex.schedulers.Schedulers;

@SuppressWarnings("unchecked")
public class MainActivity extends BaseActivity {

    private CategoryAdapter adapter;
    private ArrayList<Category> categoriesList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        disableFlags(true);
        enableWhiteStatusBar();

        setSupportActionBar(findViewById(R.id.toolbar));

        adapter = new CategoryAdapter(categoriesList);
        new EasyRecyclerView()
                .setType(EasyRecyclerView.Type.GRID)
                .setAdapter(adapter)
                .setRecyclerView(findViewById(R.id.rv_categories))
                .setItemSpacing(16, null)
                .setSpan(2)
                .initialize();

        fetchItems();
    }

    @SuppressLint("CheckResult")
    private void fetchItems() {
        ApiService service = RetrofitBuilder.createService(ApiService.class);
        StoreRepository repository = new StoreRepository(getApplication());
        repository
                .create(service.getFunctions(), new TypeToken<List<Category>>() {
                }.getType())
                .get(MyStores.CATEGORIES)
                .subscribeOn(Schedulers.io())
                .subscribe(
                        categories -> {
                            categoriesList.clear();
                            categoriesList.addAll((Collection<? extends Category>) categories);
                            runOnUiThread(() -> adapter.notifyDataSetChanged());
                        },
                        t -> ((Throwable) t).printStackTrace()
                );
    }
}

File: app/src/main/java/ga/forntoh/bableschool/MainActivity.kt
package ga.forntoh.bableschool

import android.os.Bundle
import android.support.v7.widget.Toolbar
import com.forntoh.EasyRecyclerView.EasyRecyclerView
import com.raizlabs.android.dbflow.kotlinextensions.from
import com.raizlabs.android.dbflow.kotlinextensions.list
import com.raizlabs.android.dbflow.kotlinextensions.save
import com.raizlabs.android.dbflow.kotlinextensions.select
import com.raizlabs.android.dbflow.sql.language.Delete
import ga.forntoh.bableschool.adapters.CategoryAdapter
import ga.forntoh.bableschool.model.Category
import ga.forntoh.bableschool.utils.Utils.dealWithData
import ga.forntoh.bableschool.utils.Utils.isConnected
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : BaseActivity() {

    private val adapter by lazy { CategoryAdapter(categoriesList) }
    private val categoriesList = ArrayList<Category>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        disableFlags(true)
        enableWhiteStatusBar()

        setSupportActionBar(toolbar as Toolbar?)

        EasyRecyclerView().apply {
            setType(EasyRecyclerView.Type.GRID)
            setAdapter(adapter)
            setRecyclerView(rv_categories)
            setItemSpacing(16, null)
            setSpan(2)
            initialize()
        }

        fetchItems()
    }

    private fun fetchItems() {
        val service = RetrofitBuilder.createService(ApiService::class.java)

        if (isConnected(this))
            service.functions
                    .subscribeOn(Schedulers.io())
                    .subscribe({ categories ->
                        Delete.table(Category::class.java)
                        categories.forEach { it.id = (categories.indexOf(it) + 1).toLong(); it.save() }
                        dealWithData(this, categories, categoriesList, adapter)
                    }, { dealWithData(this, (select from Category::class).list, categoriesList, adapter) })
        else dealWithData(this, (select from Category::class).list, categoriesList, adapter)
    }
}
